mm/pmm.c

```diff
@@ -310,39 +300,36 @@ static size_t process_memory_range(unsigned index, unsigned first_avail_region)
      * because initial virtual memory mapping is small.
      */
 
-    /* Add initial 4K frames and align to 2M. */
-    while ((cur < MB(EARLY_VIRT_MEM) || cur % PAGE_SIZE_2M) && cur + PAGE_SIZE <= end) {
-        if (index <= first_avail_region)
-            add_early_frame(paddr_to_mfn(cur), PAGE_ORDER_4K);
-        else
-            add_frame(paddr_to_mfn(cur), PAGE_ORDER_4K);
+    /* Add initial 4K frames for early memory. */
+    while (cur < MB(EARLY_VIRT_MEM) && (cur + PAGE_SIZE) <= end) {
+        // as long as the early frames are within the first 2M of memory, this will be
+        // fine
+        add_early_frame(paddr_to_mfn(cur), PAGE_ORDER_4K);
         cur += ORDER_TO_SIZE(PAGE_ORDER_4K);
     }
 
-    max_order = find_max_avail_order(end - cur);
-
-    /* Add all available max_order frames. */
-    while (cur + ORDER_TO_SIZE(max_order) <= end) {
-        add_frame(paddr_to_mfn(cur), max_order);
-        cur += ORDER_TO_SIZE(max_order);
-    }
+    while (cur + PAGE_SIZE <= end) {
+        unsigned int order = __builtin_ctz(cur) - PAGE_SHIFT;
```

Comment:
Nit: Please "export" the builtin function to a dedicated lib.h inline helper/wrapper (we have to start keeping all of them in a single place)

Reply:
Hmm.. How about compiler.h?